eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/DelegSpec.hs

```diff
@@ -426,6 +427,74 @@ spec = do
       expectNotRegistered cred
       expectNotDelegatedVote cred
       expectNotDelegatedToPool cred
+
+    it "Delegate, retire and re-register pool" $ do
+      expectedDeposit <- getsNES $ nesEsL . curPParamsEpochStateL . ppKeyDepositL
+      cred <- KeyHashObj <$> freshKeyHash
+      poolKh <- freshKeyHash
+      rewardAccount <- registerRewardAccount
+      registerPool poolKh
+      drepCred <- KeyHashObj <$> freshKeyHash
+
+      submitTx_ $
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . certsTxBodyL
+            .~ [ RegDepositDelegTxCert
+                  cred
+                  (DelegStakeVote poolKh (DRepCredential drepCred))
+                  expectedDeposit
+               ]
+      expectDelegatedToPool cred poolKh
+      expectDelegatedVote cred (DRepCredential drepCred)
+
+      let poolLifetime = 2
+      let poolExpiry = getsNES nesELL <&> \n -> addEpochInterval n $ EpochInterval poolLifetime
+
+      poolExpiry >>= \pe ->
+        submitTx_ $
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . certsTxBodyL .~ [RetirePoolTxCert poolKh pe]
+
+      -- when pool is re-registered after its expiration, all delegations are cleared
+      passNEpochs $ fromIntegral poolLifetime
+      expectNotDelegatedToPool cred
+      registerPoolWithRewardAccount poolKh rewardAccount
+      expectNotDelegatedToPool cred
+      -- the vote delegation is kept
+      expectDelegatedVote cred (DRepCredential drepCred)
+
+      -- re-delegate
+      submitTx_ $
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . certsTxBodyL
+            .~ [ DelegTxCert
+                  cred
+                  (DelegStake poolKh)
+               ]
+      expectDelegatedToPool cred poolKh
+
+      -- when pool is re-registered before its expiration, delegations are kept
+      poolExpiry >>= \pe ->
+        submitTx_ $
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . certsTxBodyL .~ [RetirePoolTxCert poolKh pe]
+      -- re-register the pool before the expiration time
+      passNEpochs $ fromIntegral poolLifetime - 1
+      registerPoolWithRewardAccount poolKh rewardAccount
+      expectDelegatedToPool cred poolKh
+      passNEpochs 2
+      expectDelegatedToPool cred poolKh
+
+      -- when pool is retired and re-registered in the same transaction, delegations are kept
```

Comment:
Don't know if it's better to have all these scenarios in separate tests . On one hand,  in separate tests, it's easier to see which fails and to isolate the cause in case of failure, on the other hand, it would produce more noise and repetition.
I don't have an inclination to either,  I'm happy to change to separate tests if that's more idiomatic.

Reply:
I feel exactly the same way :smile: